atlassian/bitbucket/cloud/repositories/pullRequests.py

```diff
@@ -0,0 +1,169 @@
+
+    @property
+    def state(self):
+        """
+        pull request state
+        possible values: MERGED, SUPERSEDED, OPEN, DECLINED
+        """
+        return self.get_data("state")
+
+    @property
+    def created_on(self):
+        """ time of creation """
+        return datetime.strptime(self.get_data("created_on"), "%Y-%m-%dT%H:%M:%S.%f%z")
+
+    @property
+    def updated_on(self):
+        """ time of last update """
+        uo_str = self.get_data("updated_on")
+        uo_dt = datetime.strptime(uo_str, "%Y-%m-%dT%H:%M:%S.%f%z") if uo_str else uo_str
+        return uo_dt
+
+    @property
+    def close_source_branch(self):
+        """ close source branch flag """
+        return self.get_data("close_source_branch")
+
+    @property
+    def source_branch(self):
+        """ source branch """
+        branch = self.get_data("source").get("branch")
+        return branch.get("name")
+
+    @property
+    def destination_branch(self):
+        """ destination branch """
+        branch = self.get_data("destination").get("branch")
+        return branch.get("name")
+
+    @property
+    def comment_count(self):
+        """ number of comments """
+        return self.get_data("comment_count")
+
+    @property
+    def task_count(self):
+        """ number of tasks """
+        return self.get_data("task_count")
+
+    @property
+    def declined_reason(self):
+        """ reason for declining """
+        return self.get_data("reason")
+
+    @property
+    def author(self):
+        """ DefaultReviewer object of the author """
+        return DefaultReviewer(None, self.get_data("author"))
+
+    def participants(self):
+        """ Returns a generator object of participants """
+        for participant in self.get_data("participants"):
+            yield Participant(participant)
+
+        return
+
+    def reviewers(self):
+        """ Returns a generator object of reviewers """
+        for reviewer in self.get_data("reviewers"):
+            yield DefaultReviewer(None, reviewer)
+
+        return
+
+
+class Participant(BitbucketCloudBase):
+    def __init__(self, data, *args, **kwargs):
+        super(Participant, self).__init__(None, None, *args, data=data, expected_type="participant", **kwargs)
+
+    @property
+    def user(self):
+        """ DefaultReviewer object with user information of the participant """
+        return DefaultReviewer(None, self.get_data("user"))
+
+    @property
+    def role(self):
+        """ Returns PARTICIPANT or REVIEWER """
+        return self.get_data("role")
+
+    @property
+    def state(self):
+        """ Returns approved, changes_requested or None"""
+        return self.get_data("state")
+
+    @property
+    def participated_on(self):
+        """ time of last participation """
+        po_str = self.get_data("participated_on")
+        po_dt = datetime.strptime(po_str, "%Y-%m-%dT%H:%M:%S.%f%z") if po_str else po_str
+        return po_dt
```

Comment:
I thought returning a `datetime` object is more comfortable. In the most cases you would do this convention yourself later. 

[It like implementing `state` vs `is_open`, `is_closed` etc. ](https://github.com/atlassian-api/atlassian-python-api/pull/655#discussion_r546903402)

Reply:
We should do this on the other classes as well

Reply:
Ok. The already existing objects return the raw string and I don't like the hardcoded format. What about using a `timeformat` in the constructor with a default of None. The user can specify a format in the main constructor:
```python
cloud = Cloud(url="http://localhost:7990", username="admin", password="admin", timeformat="%Y-%m-%dT%H:%M:%S.%f%z")
```
The format must be given to the sub classes by extending the following function with the `timeformat`.
https://github.com/atlassian-api/atlassian-python-api/blob/eec2934fcd72ebf35205296bf2ddb3f94680c0ed/atlassian/bitbucket/base.py#L66-L72
In the base class we can define the function `get_time`:
```python
def get_time(self, id, default):
    if id not in self.__data:
        return default
   timestring = self.__data[id]
    if self.timeformat is not None:
        timestring = datetime.strptime(timestring, self.timeformat)
    return timestring
```

The function can than be used the smae way as `get_data`:
```python
    @property
    def completed_on(self):
        return self.get_time("completed_on", "never completed"))
```

Or maybe the `get_data`

Reply:
But `%Y-%m-%dT%H:%M:%S.%f%z` is the delivered Timeformat from BB.  It's necessary to use this format to convert `string` to `datetime`. Therefore it doesn't make sense to use custom set time format by the user. We can add a parameter to return raw timestamp or `datetime` object.

Reply:
Ok. My understanding was wrong. I thougt the function interprets the string and formats it according to the argument. The parameter to return the raw time is a good idea. Can this be combined with my wrong understanding of the function to give a format? E.g. Add a `timeformat` argument with a default of `""` which returns the `datetime` object, a value of `None` returns the raw string and other values are given to the function `strftime`of the `datetime` object.

Reply:
Yes, but i suggest only to impement a `raw_timeformat` boolean with default `False`. Convertings are user based. Most cases you need to convert to your timezone because BB send UTC time etc.

Reply:
That's why I suggested a format string in the constructor. Than the user can define the format in a single place and don't have to take care about converting the time properties.

Reply:
But if you like to get a time string `DD.MM.YYYY HH24:MI` in GMT+2 you need to convert the timezone of the `datetime` object first and afterwards you could parse it to a string with the given format. Therefore you need a min. of two parameters (time format + timezone).

Reply:
Added the possibility to define a `raw_timezone` boolean (default `False`). Maybe we could reconsider the suggestion about a pre-formatted time-string later. IMHO thats no breaking change on a later implementation.

Reply:
Later this will be a braking change or an additional argument is needed. What about configuring a function to change the time? None will return the raw string and the default `lamda x: x` will return the datetime object?

Reply:
Good catch. Added this in the last commit.

Default is convert to UTC `datetime`, if passing `None` returns the raw string. Otherwise you are able to pass a `lambda` function in.

```
import pendulum
from datetime import datetime
from atlassian.bitbucket.cloud import Cloud
from atlassian.bitbucket.cloud.const import CONF_TIMEFORMAT

def timeformat(x):
    tz = pendulum.timezone("Europe/Berlin")
    x = datetime.strptime(x, CONF_TIMEFORMAT)
    return tz.convert(x)


bb = Cloud(
    url="https://api.bitbucket.org",
    username="xxxx",
    password="xxxx",
    timeformat_func=lambda x: timeformat(x), # returns converted datetime object in timezone Europe/Berlin
    #timeformat_func=None, # -> returns raw string
)
```

Reply:
@Spacetown what about the parameter name? Should it be `timeformat_func` or `timeformat_lambda` or do you have another idea.

I tend to `timeformat_lambda`

Reply:
I think `timeformat_lambda` is fine.

Reply:
I think the line `x = datetime.strptime(x, CONF_TIMEFORMAT)` should be done in `get_time`:
```python
    def get_time(self, id):
        value_str = self.get_data(id)
        if not value_str or not isinstance(value_str, str):
            return None
        else:
            return value_str if self.timeformat_func is None else self.timeformat_func(datetime.strptime(x, CONF_TIMEFORMAT))
```
Than the user doesn't need to know the time formet `CONF_TIMEFORMAT` and if the user need to format the time he always need the parsed string.

Reply:
Please check the new implementation. I added a default lambda function which handles the string checks. But you are able now to pass in another lambda function to e.g. return a default value if `None`. Something like this:

```
from atlassian.bitbucket.cloud import Cloud
from atlassian.bitbucket.base import CONF_TIMEFORMAT
from datetime import datetime
import pendulum


def timeformat(x):
    tz = pendulum.timezone("Europe/Berlin")
    if isinstance(x, str):
        x = datetime.strptime(x, CONF_TIMEFORMAT)
        return tz.convert(x)
    else:
        return "no date"


bb = Cloud(
    url="https://api.bitbucket.org",
    username="xxx",       
    password="xxxx",
    timeformat_lambda=lambda x: timeformat(x),
    # timeformat_lambda=None,
)
```

But you still have to import the constant. But I think this implementation is more flexibel for the user. Passing the date convert directly into the lambda would not work with that approach because `strptime` expects a `str`.

Another approach would be:
```
 def get_time(self, id):
        value_str = self.get_data(id)
        if isinstance(value_str, str):
            value = datetime.strptime(value_str, CONF_TIMEFORMAT)
        else:
            value = value_str

        return value_str if self.timeformat_lambda is None else self.timeformat_lambda(value)
```
This would pass the converted `datetime` object or the pure type (if not `str`). Everything else must be handled in the lambda function. In that scenario, the import of `CONF_TIMEFORMAT` is not necessary.

Reply:
I think the second approach is Vetter because the User doesn't need to take Cafe about the interna format. Getting the local time is quite easy.
```python
def timeformat(x):
    if x:
        return pendulum.timezone("Europe/Berlin").convert(x)
    else:
        return "no date"
```
Converting the time into an object in `get_time` is only needed if a lambda is defined. If the User requests raw strings the conversion isn't needed.
```python
 def get_time(self, id):
        value_str = self.get_data(id)
        if self.timeformat_lambda is None:
            return value_str

        if isinstance(value_str, str):
            value = datetime.strptime(value_str, CONF_TIMEFORMAT)
        else:
            value = value_str

        return self.timeformat_lambda(value)
```

Reply:
Added this approach.